roadmap/state-sync-protocol.md

```diff
@@ -0,0 +1,198 @@
+```
+
+**Execution Data (individual)**
+
+```console
+smnzhu@mbp:~$ SORTED_BY_SIZE=$(gsutil du -h gs://flow_public_mainnet13_execution_state | sed -e 's/ MiB/M/g' -e 's/ KiB/K/g' | sort -h)
+smnzhu@mbp:~$ echo $SORTED_BY_SIZE | head -n 1
+3.57K     gs://flow_public_mainnet13_execution_state/5d00b1ef2670d5987015c582727ab2308b19916ea4303ad56d600bb1fb4a12fc.cbor
+smnzhu@mbp:~$ echo $SORTED_BY_SIZE | tail -n 1
+377.18M   gs://flow_public_mainnet13_execution_state/38db1ad86322fa81ff105328fd2812f2d5eed87ea4b7dba4e4095b1934fcec86.cbor
+smnzhu@mbp:~$ echo $SORTED_BY_SIZE | awk '{ a[i++]=$0; } END { print a[int(i/2)]; }'
+70.47K    gs://flow_public_mainnet13_execution_state/4f6892c5b1b16e41ae1adda086b8aa241d254b780f44ceb5e45a5a1d5bb3d8aa.cbor
+```
+
+**Execution Data (~3 week spork)**
+
+```console
+smnzhu@mbp:~$ gsutil du -sh gs://flow_public_mainnet13_execution_state
+446.14 GiB   gs://flow_public_mainnet13_execution_state
+smnzhu@mbp:~$ gsutil ls gs://flow_public_mainnet13_execution_state | wc -l
+  473590
+```
+
+**Execution State Index (~3 week spork)**
+
+```console
+simon_zhu_dapperlabs_com@observer-001 /var/flow/data/mainnet-13 $ sudo du -sh index
+103G    index
+```
+
+Based on the numbers above, we should require Access nodes to have >43GB of RAM (27GB + current 16GB requirement) and minimum 1.5-2TB of storage space for a 3 month spork. These requirements can likely be decreased if sharding is implemented (see [Future Optimizations](#future-optimizations)).
+
+## Future Optimizations
+
+### Checkpointing
+
+Having a checkpointing mechanism will allow new nodes to catch up more quickly, while also reducing the amount of data Access nodes need to store on disk.
+
+Access nodes can periodically (e.g. at every 10000th block) generate a snapshot of the full state index and store it in their blobstore. They can then prune all of the Execution Data for blocks older than the snapshot, and require nodes that are behind to directly sync the snapshot first before catching up via per-block Execution Data updates.
+
+The root CID of the snapshot must be shared with new nodes somehow so that they can use it to download the snapshot via Bitswap. There are a few ways this can be done:
+
+* The CID can be shared on a broadcast channel using a *Byzantine consistent broadcast* algorithm<sup>[5](#footnotes)</sup>.
+* A new Access node can randomly select a couple of other Access nodes and simply ask them for the CID. If the responses it receives are concordant, it can go ahead and download the data. Once the download is complete, it can generate the state trie for each block and compare its hash to the one found on-chain in the corresponding Execution Result. If it finds an inconsistency, it has all the evidence it needs to submit a slashing challenge against the nodes which provided it with the bad CID.
+
+### Sharding
+
+The Execution Data can be sharded so that each Access node only needs to store and respond to requests for a fraction of the entire dataset.
```

Comment:
somewhat similar to collection nodes and their clustering.

but yes, as mentioned later, this increased the severity of possible data availability issues, and should likely only be pursued after a critical mass of access nodes are part of the network